### ISSUE TITLE
linear regression issues

### ISSUE THREAD

**Issue by Len-Palmeri** [Tom-Alexander/regression-js#18]
Thank you for sharing your fine work on regression-js.  I just wanted to point out 2 issues that caused my linear best-fit to miscompute the gradient: 
1. after pushing my actual data points onto bfSeries (see MAIN excerpt below), numeric coords were sometimes being interpreted as _string_ within the regression.js linear function for some unknown reason, throwing off the gradient computation.  To prevent this, I call parseFloat() before pushing any of the points.
2. after pushing my 2 trend-setting points  (having null y-coord) onto bfSeries, the linear function in regression.js was trying to use these points when computing gradient & intercept. To avoid this, I introduced var N (see below).
### MAIN JAVASCRIPT:

``` javascript
<script type="text/javascript" src="/includes/regression.js"></script>
var bfSeries = []; //define empty array to hold each actual data point plus 2 trend-setting points.

//Prepare a data series for Best Fit (bfSeries), based on an existing series of actual data points.
bfSeries.push([parseFloat(1), null]);   //First push the left trend-setting point; the null tells regression() to compute the y coordinate of this point on the trend line.
for (i=0; i < series.length; i++) {    //push each point of actual data. Only the actual data points are used to compute the trend line.               
    bfSeries.push([parseFloat(series[i][0]), parseFloat(series[i][1])]);   //Bug: string coords were observed to be pushed here, causing the regression to miscompute slope.  Fix: convert coords to float prior to pushing.
}
bfSeries.push([parseFloat(100), null]);  //Finally, push the right trend-setting point; the null tells regression() to compute the y coordinate of this point on the trend line.
var myRegression = regression('linear', bfSeries);
```
### REGRESSION.JS (excerpt):     I introduce var N to fix forecasting issue 2:

``` javascript
            linear: function(data) {
                var sum = [0, 0, 0, 0, 0], n = 0, results = [];
                var N = 0;  //the actual number of x,y points participating in the regression (excludes forecasted points, i.e., those having a null y-coordinate). 

                for (; n < data.length; n++) {
                  if (data[n][1]) {
                    sum[0] += data[n][0];
                    sum[1] += data[n][1];
                    sum[2] += data[n][0] * data[n][0];
                    sum[3] += data[n][0] * data[n][1];
                    sum[4] += data[n][1] * data[n][1];
                    N++;
                  }
                }

             // var gradient = (n * sum[3] - sum[0] * sum[1]) / (n * sum[2] - sum[0] * sum[0]);
             // var intercept = (sum[1] / n) - (gradient * sum[0]) / n;
             // var correlation = (n * sum[3] - sum[0] * sum[1]) / Math.sqrt((n * sum[2] - sum[0] * sum[0]) * (n * sum[4] - sum[1] * sum[1]));

                var gradient = (N * sum[3] - sum[0] * sum[1]) / (N * sum[2] - sum[0] * sum[0]);
                var intercept = (sum[1] / N) - (gradient * sum[0]) / N;

                for (var i = 0, len = data.length; i < len; i++) {
                    var coordinate = [data[i][0], data[i][0] * gradient + intercept];  //this plots all data points (including forcasted points)
                    results.push(coordinate);
                }

                var string = 'y = ' + Math.round(gradient*100) / 100 + 'x + ' + Math.round(intercept*100) / 100;
             // var string = 'y = ' + gradient + 'x + ' + intercept;   //actual equation
                return {equation: [gradient, intercept], points: results, string: string};
            },
```

**Comment by joneldiablo** [comment]
when all the x values are the same, the NaN error is coming again!!!! for example, regression('linear', [[parseFloat(1),parseFloat(1)],[parseFloat(1),parseFloat(1)],[parseFloat(1),parseFloat(1)],[parseFloat(1),parseFloat(1)],[parseFloat(1),parseFloat(1)],[parseFloat(01),parseFloat(1)]]);

**Comment by Tom-Alexander** [comment]
Fixed in #59
